### PR TITLE
Pass user session information when opening the electron app from web platform

### PR DESCRIPTION
**Pull request by ntdiary**
### Fixed Issues

$ https://github.com/Expensify/App/issues/15651
PROPOSAL: https://github.com/Expensify/App/issues/15651#issuecomment-1467954112

### Tests

1. Login with account A in the desktop app.
2. Login with account B in the web app.
3. Navigate to the user profile page in the web app.
4. Refresh the browser.
5. A prompt should appear asking to 'Open the app in Electron'.
6. Click on it and wait for the desktop app to open.
7. Verify the user profile page for account B is displayed in the desktop app.

- [x] Verify that no errors appear in the JS console

### Offline tests

1. Login with account A in the desktop app.
2. Login with account B in the web app.
3. Navigate to the user profile page in the web app.
4. Refresh the browser.
5. A prompt should appear asking to 'Open the app in Electron'.
6. Turn off the network.
7. Click on 'Open the app in Electron' and wait for the desktop app to open.
8. Verify the login page is displayed in the desktop app.

### QA Steps

1. Login with account A in the desktop app.
2. Login with account B in the web app.
3. Navigate to the user profile page in the web app.
4. Refresh the browser.
5. A prompt should appear asking to 'Open the app in Electron'.
6. Click on it and wait for the desktop app to open.
7. Verify the user profile page for account B is displayed in the desktop app.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/8579651/225961641-73dadfd3-71a4-431a-ba76-8d0bd191820e.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>
N/A

</details>

<details>
<summary>Mobile Web - Safari</summary>
N/A

</details>

<details>
<summary>Desktop</summary>
N/A

</details>

<details>
<summary>iOS</summary>
N/A

</details>

<details>
<summary>Android</summary>
N/A

</details>

### CODE REVIEW

**Comment by hayata-suenaga** [comment]
@ntdiary I'm looking at your questions right now. I'll get back to you after I get insights into these.

**Comment by hayata-suenaga** [comment]
@ntdiary FYI the desktop deeplink [is not working on main currently](https://github.com/Expensify/App/issues/15965).

File: src/components/DeeplinkWrapper/index.website.js
Comment (hayata-suenaga) [reply src/components/DeeplinkWrapper/index.website.js:68]: We have [custom URL scheme.](https://github.com/Expensify/App/blob/0efe6acf41eb985a0d6e42b47a8d91ab9cef6559/src/CONST.js#L272) `new-expensify://`.

I think somewhere in the code, if the desktop is registering that URL schema with Mac OS, telling it that if OS detects user opening any URL that starts with `new-expensify://`, please let the desktop app handle what to do with the URL.

**Comment by hayata-suenaga** [comment]
I'm looking at your [comment](https://github.com/Expensify/App/pull/16043#issuecomment-1475817110) now I have to be offline now and I'm OOO today, but if I can find a time tonight, I'll get back to you.

**Comment by hayata-suenaga** [comment]
@ntdiary thank you for checking on auth token lifespan.

> I temporarily let the app jump to the home page instead of getting stuck on the /transition page.

Is it possible for us to intercept the event of clicking on `Open Electron` button and navigate the user back to the home page?

**Comment by hayata-suenaga** [comment]
@ntdiary does the user have to sign in again? not just refreshing the screen?

**Comment by hayata-suenaga** [comment]
I'll wait @shawnborton's response but shouldn't we display "Please refresh again" instead?

And @ntdiary how do you suggest we trigger the change of the screen when the token expires?

**Review by hayata-suenaga — DISMISSED** [review]
ohhh I completely misunderstood. yes it makes sense to prompt the user to log in again on the desktop side. let's wait for @shawnborton's review but otherwise LGTM!

**Comment by hayata-suenaga** [comment]
there is a suggestion that we should implement a new screen for expired session and deprecate `DeeplinkWrapper`. @ntdiary please let me know if you're up to this 👀

https://expensify.slack.com/archives/C049HHMV9SM/p1679504172500439?thread_ts=1678823496.930569&cid=C049HHMV9SM
https://expensify.slack.com/archives/C049HHMV9SM/p1679505364500279?thread_ts=1678823496.930569&cid=C049HHMV9SM

**Comment by hayata-suenaga** [comment]
>  LogInWithShortLivedAuthTokenPage (not yet committed)

@ntdiary I'll wait for your commit in the expired session screen.

File: src/languages/es.js
Comment (hayata-suenaga) [inline src/languages/es.js:160]: did you check for the translation on slack?

File: src/ONYXKEYS.js
Comment (hayata-suenaga) [inline src/ONYXKEYS.js:199]: should we use `isTokenValid` instead?

File: src/languages/es.js
Comment (pecanoro) [inline src/languages/es.js:160]:
```suggestion
        signIn: 'Por favor, inicia sesión de nuevo.',
```

File: src/languages/es.js
Comment (pecanoro) [inline src/languages/es.js:159]:
```suggestion
        expired: 'Tu sesión ha expirado.',
```

**Review by pecanoro — CHANGES_REQUESTED** [review]

**Review by thesahindia — CHANGES_REQUESTED** [review]
Looks good! Just a minor change. 

I will test and fill the checklist in a few minutes

**Review by thesahindia — DISMISSED** [review]

File: src/pages/LogInWithShortLivedAuthTokenPage.js
Comment (pecanoro) [inline src/pages/LogInWithShortLivedAuthTokenPage.js:38]: All props need a comment explaining what they are for

**Review by pecanoro — CHANGES_REQUESTED** [review]

**Review by thesahindia — DISMISSED** [review]

File: src/libs/actions/Session/index.js
Comment (hayata-suenaga) [inline src/libs/actions/Session/index.js:275]: should we update the comments here since we're also dealing with the situation of transitioning from web NewDot to desktop NewDot?

File: src/components/DeeplinkWrapper/index.website.js
Comment (pecanoro) [inline src/components/DeeplinkWrapper/index.website.js:85]: Why catch and not always?

**Comment by hayata-suenaga** [comment]
@ntdiary I'm waiting for @thesahindia's review.

**Review by thesahindia — APPROVED** [review]
It looks good to me @hayata-suenaga

**Review by hayata-suenaga — APPROVED** [review]
LGTM!

**Review by pecanoro — APPROVED** [review]

**Review by tgolen — COMMENTED** [review]
Hi, I'm sorry to be coming into this late, but someone asked me about some of the code implemented in this PR and I had many questions about it.

**Comment by hayata-suenaga** [comment]
@tgolen thank you very much for the comments. we should have done more thorough review.

Some of the fixes suggested changed are being implemented in this [PR](https://github.com/Expensify/App/pull/17364)

@ntdiary are you up for a followup PR for fixing the original one?

**Comment by tgolen** [comment]
Great, thank you so much for addressing these! It's gonna make this code a lot cleaner. 👍

**Comment by s77rt** [comment]
I think this PR caused a regression https://github.com/Expensify/App/issues/17391 (details [here](https://github.com/Expensify/App/issues/17391#issuecomment-1512464169)). This PR adds the transition prefix to all deeplinks and I think we should not do that because the transition prefix is only for OLDDOT -> NEWDOT and not WEB -> DESKTOP.

**Pull request merged**
